File: src/org/elixir_lang/debugger/Modules.java
package org.elixir_lang.debugger;

import org.elixir_lang.jps.builder.ParametersList;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;

public class Modules {
    private static final String BASE_PATH = "/debugger";
    private static final String INTELLIJ_ELIXIR_DEBUG_SERVER = "lib/debug_server.ex";
    private static final String MIX_TASKS_INTELLIJ_ELIXIR_DEBUG_TASK = "lib/debug_task.ex";
    private static final List<String> RELATIVE_SOURCE_PATH_LIST = Arrays.asList(
            INTELLIJ_ELIXIR_DEBUG_SERVER,
            MIX_TASKS_INTELLIJ_ELIXIR_DEBUG_TASK
    );

    private Modules() {
    }

    @NotNull
    public static ParametersList add(@NotNull ParametersList parametersList) throws IOException {
        return org.elixir_lang.ElixirModules.add(parametersList, copy());
    }

    @NotNull
    private static List<File> copy() throws IOException {
        return org.elixir_lang.ElixirModules.copy(BASE_PATH, RELATIVE_SOURCE_PATH_LIST);
    }
}

File: src/org/elixir_lang/debugger/Modules.kt
package org.elixir_lang.debugger

import java.io.File
import java.io.IOException

object Modules {
    private const val BASE_PATH = "/debugger"
    private const val INTELLIJ_ELIXIR_DEBUGGER_SERVER = "lib/intellij_elixir/debugger/server.ex"
    private const val INTELLIJ_ELIXIR_DEBUGGED = "lib/intellij_elixir/debugged.ex"
    private const val MIX_TASKS_INTELLIJ_ELIXIR_DEBUG = "lib/mix/tasks/intellij_elixir/debug.ex"
    private val RELATIVE_SOURCE_PATH_LIST = listOf(
            INTELLIJ_ELIXIR_DEBUGGER_SERVER,
            INTELLIJ_ELIXIR_DEBUGGED
    )

    @JvmStatic
    @Throws(IOException::class)
    fun add(parametersList: MutableList<String>): List<String> =
            org.elixir_lang.ElixirModules.add(parametersList, copy())

    private fun copy(mix: Boolean = false): List<File> =
            org.elixir_lang.ElixirModules.copy(BASE_PATH, relativeSourcePathList(mix))

    private fun relativeSourcePathList(mix: Boolean): List<String> =
            RELATIVE_SOURCE_PATH_LIST +
                    if (mix) {
                        listOf(MIX_TASKS_INTELLIJ_ELIXIR_DEBUG)
                    } else {
                        emptyList()
                    }

    fun erlArgumentList(mix: Boolean = false): List<String> =
            listOf("-eval", "'application:ensure_all_started(elixir)'") +
                    copy(mix).flatMap { file ->
                        listOf("-eval", "'Elixir.Code:require_file(\"${file.path}\")'")
                    } + if (mix) {
                emptyList()
            } else  {
                listOf("-eval", "'Elixir.IntelliJElixir.Debugged:start()'")
            }
}
